### PR TITLE
Macro joining Thrift structs.

### PR DESCRIPTION
**Pull request by huragan**

### CODE REVIEW

**Comment by SamRoberts** [comment]
The macro now fails compilation upon encountering an input field with the same name as an output field but an incompatible type.

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/transform/Transform.scala [hunk 1]
```diff
@@ -15,3 +15,4 @@
 package au.com.cba.omnia.maestro.core.transform
 
 case class Transform[A, B](run: A => B)
+case class Join[A, B](run: A => B)
```
Comment (stephanh) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/transform/Transform.scala]: We should probably add some scaladoc for these. In particular we should document that `A` is a product.

File: maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+package au.com.cba.omnia.maestro.macros
+
+import scala.reflect.macros.Context
+
+import scalaz._, scalaz.syntax.either._
+
+import com.twitter.scrooge.ThriftStruct
+
+import au.com.cba.omnia.humbug.HumbugThriftStruct
+import au.com.cba.omnia.maestro.core.transform.Join
```
Comment (stephanh) [inline maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala]: There should be a new line here.

File: maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+/**
+  * Macro to automatically derive transformations from a tuple of thrift struct
```
Comment (stephanh) [inline maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala]: thrift structs (plural).

File: maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+  * The macro will generate a `Join` value which matches input fields to
+  * output fields if they have the same name and compatible types. If multiple
+  * input fields match, they must have the same values at runtime or the `Join`
+  * value will throw an exception.
+  *
+  * Compilation will succeed when:
+  * - The input type is a product (e.g. tuple) of thrift structures.
+  * - Every field in the output thrift struct has matching fields (fields with
+  *   the same name) amongst the input thrift structs.
+  * - Every matching input field's type is assignable to the type of it's
```
Comment (stephanh) [inline maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala]: Assignable or the same?

File: maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+object JoinMacro {
+  def impl[A <: Product : c.WeakTypeTag, B <: ThriftStruct : c.WeakTypeTag](c: Context)
+      : c.Expr[Join[A, B]] = {
+    import c.universe._
+
+    // thrift types and fields
+    case class ThriftField(name: TermName, typ: Type)
+    case class ThriftType(typ: Type, fields: List[ThriftField])
+
+    /** A thrift type inside the input product. name is the name of the input product field */
```
Comment (stephanh) [inline maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala]: No need for scaladoc comments here. Simple `/* */` will do.

File: maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+
+    val tree      = for {
+      inputs     <- thriftInputs(inputType)
+      mappings   <- mappingsFor(output.fields, inputs)
+    } yield joinTree(inputType, output, inputs, mappings)
+
+    c.Expr[Join[A, B]](tree.fold(msg => abort(msg), identity))
+  }
+
+  type Fails[A] = String \/ A
```
Comment (stephanh) [inline maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala]: Should probably go to the top of the method.

File: maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+          if (typ <:< weakTypeOf[ThriftStruct])
+            ThriftInput(newTermName(sym.name.toString.trim), thriftType(typ)).right
+          else
+            s"$typ is not a thrift struct".left
+        }
+      }
+      collectErrors(thriftInputs, errs => s"Join requires a product of thrift structs: ${errs.mkString(", ")}")
+    }
+
+    /** get the mappings from input fields to output fields,
```
Comment (stephanh) [inline maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala]: Wrong commenting style.

File: maestro-macros/src/test/scala/au/com/cba/omnia/maestro/macros/JoinMacroSpec.scala [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    """import au.com.cba.omnia.maestro.macros._
+       import au.com.cba.omnia.maestro.test.thrift.humbug._
+    """
+
+  val importsScrooge =
+    """import au.com.cba.omnia.maestro.macros._
+       import au.com.cba.omnia.maestro.test.thrift.scrooge._
+    """
+
+  val scroogeStructOne = au.com.cba.omnia.maestro.test.thrift.scrooge.JoinOne("some-string", true, 42)
```
Comment (stephanh) [inline maestro-macros/src/test/scala/au/com/cba/omnia/maestro/macros/JoinMacroSpec.scala]: Can we make these top level imports?

File: maestro-macros/src/test/scala/au/com/cba/omnia/maestro/macros/JoinMacroSpec.scala [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+
+  def incompatibleTypes = {
+    lazy val compileMacro = {
+      new Eval()(importsScrooge + """Macros.mkJoin[(JoinOne, JoinTwo, JoinOneIncompatible), Joinable]""")
+    }.toString
+
+    compileMacro must throwA[com.twitter.util.Eval$CompilerException].like {
+      case e => e.getMessage must contain("someField is not assignable from")
+    }
+
```
Comment (stephanh) [inline maestro-macros/src/test/scala/au/com/cba/omnia/maestro/macros/JoinMacroSpec.scala]: Superfluous new line.

**Comment by stephanh** [comment]
:+1: just needs some cosmetic changes.

File: maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/transform/Transform.scala (hunk 1 above)
Comment (SamRoberts) [inline maestro-core/src/main/scala/au/com/cba/omnia/maestro/core/transform/Transform.scala]: Very good point

File: maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala (hunk 4 above)
Comment (SamRoberts) [inline maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala]: `<:<`. Which according to scala's docs just means "conforms"

File: maestro-macros/src/test/scala/au/com/cba/omnia/maestro/macros/JoinMacroSpec.scala (hunk 8 above)
Comment (SamRoberts) [inline maestro-macros/src/test/scala/au/com/cba/omnia/maestro/macros/JoinMacroSpec.scala]: Not directly, as we use scrooge and humbug types with the same name. We could change the original types names or rename them during import though. I think I'll change the original names.

File: maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala (hunk 4 above)
Comment (SamRoberts) [inline maestro-macros/src/main/scala/au/com/cba/omnia/maestro/macros/JoinMacro.scala]: The full definition of conforms, taking into account all of the different type constructs in scala, is pretty huge, so I think I'll just use the word "conform" itself. It's both intuitively and technically correct.

**Pull request merged**